modules/sdk-coin-sui/src/lib/transaction.ts

```diff
@@ -1,16 +1,360 @@
+      return writer.writeVec(bytes, (writer, el) => writer.write8(el));
+    },
+    (reader) => {
+      const bytes = reader.readVec((reader) => reader.read8());
+      return encodeStr(new Uint8Array(bytes), 'base64');
+    }
+  );
+
+bcs.registerStructType('SuiObjectRef', {
+  objectId: 'ObjectID',
+  version: 'u64',
+  digest: 'ObjectDigest',
+});
+
+bcs
+  .registerVectorType('vector<SuiAddress>', 'SuiAddress')
+  .registerVectorType('vector<SuiObjectRef>', 'SuiObjectRef')
+  .registerStructType('PayTx', {
+    coins: 'vector<SuiObjectRef>',
+    recipients: 'vector<SuiAddress>',
+    amounts: 'vector<u64>',
+  });
+
+bcs.registerEnumType('Transaction', {
+  TransferObject: 'TransferObjectTx',
+  Publish: 'PublishTx',
+  Call: 'MoveCallTx',
+  TransferSui: 'TransferSuiTx',
+  Pay: 'PayTx',
+});
+
+bcs.registerVectorType('vector<Transaction>', 'Transaction').registerEnumType('TransactionKind', {
+  Single: 'Transaction',
+  Batch: 'vector<Transaction>',
+});
+
+bcs.registerStructType('TransactionData', {
+  kind: 'TransactionKind',
+  sender: 'SuiAddress',
+  gasPayment: 'SuiObjectRef',
+  gasPrice: 'u64',
+  gasBudget: 'u64',
+});
+
+bcs.registerType(
+  'SignatureData',
+  (writer, str) => {
+    const bytes = Array.from(decodeStr(str, 'base64'));
+    return writer.writeVec(bytes, (writer, el) => writer.write8(el));
+  },
+  (reader) => {
+    const bytes = reader.readVec((reader) => reader.read8());
+    return encodeStr(new Uint8Array(bytes), 'base64');
+  }
+);
+
+bcs.registerEnumType('Signature', {
+  Ed25519SuiSignature: 'SignatureData',
+  Secp256k1SuiSignature: 'SignatureData',
+});
+
+bcs.registerStructType('SenderSignedData', {
+  data: 'TransactionData',
+  txSignature: 'vector<u8>',
+});
+
+const UNAVAILABLE_TEXT = 'UNAVAILABLE';
+export const TYPE_TAG = Array.from('TransactionData::').map((e) => e.charCodeAt(0));
+
+export type SuiObjectRef = {
+  /** Hex code as string representing the object id */
+  objectId: string;
+  /** Object version */
+  version: number;
+  /** Base64 string representing the object digest */
+  digest: string;
+};
+
+export type SuiTransaction = {
+  sender: string;
+  kind: PayTransactionKind;
+  gasBudget: number;
+  gasPrice: number;
+  gasPayment?: SuiObjectRef;
+};
 
 export class Transaction extends BaseTransaction {
+  private _suiTransaction: SuiTransaction;
+  private _signature;
+  private _publicKey;
+
+  constructor(_coinConfig: Readonly<CoinConfig>) {
+    super(_coinConfig);
+  }
+
+  get suiTransaction(): SuiTransaction {
+    return this._suiTransaction;
+  }
+
+  setSuiTransaction(value: SuiTransaction): void {
+    this._suiTransaction = value;
+  }
+
+  /** @inheritDoc **/
+  get id(): string {
+    // Solana transaction ID === first signature: https://docs.solana.com/terminology#transaction-id
+    if (this._signature) {
+      return base58.encode(this._signature);
+    } else {
+      return UNAVAILABLE_TEXT;
+    }
+  }
```

Comment:
this needs to be the digest instead of the signature.

example for how we can calculate it: https://github.com/BitGo/coins-sandbox/pull/209/files

for now, let's just return `UNAVAILABLE_TEXT` and create a ticket to come back to it.

Reply:
with this, we can remove `bs58` from `package.json`

Reply:
oh i see that you set `this._id` further below.

you can just return that (or `UNAVAILABLE_TEXT` if it doesnt exist)

Reply:
updated